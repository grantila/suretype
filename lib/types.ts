export type ExportRefMethod =
	| 'no-refs'  // Don't ref anything. Inline all types to monolith types.
	| 'provided' // Reference types that are explicitly provided.
	| 'ref-all'; // Ref all provided types and those with names, suretype()'d.

export type OnTopLevelNameConflict =
	| 'error'   // Fail the operation
	| 'rename'; // Rename the validators to a unique name

export type OnNonSuretypeValidator =
	| 'error'       // Fail the operation
	| 'ignore'      // Ignore, don't export
	| 'create-name' // Create a name 'Unknown'
	| 'lookup';     // Provide in lookup table
