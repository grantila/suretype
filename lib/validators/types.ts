
export type Type =
	| "string"
	| "number"
	| "integer"
	| "object"
	| "array"
	| "boolean"
	| "null";

export type AnyType =
	| Type
	| "raw"
	| "any"
	| "any-of"
	| "all-of"
	| "if"
	| "recursive";

export type FilterProperties< T, Cond > =
	{
		[ K in keyof T ]: T[ K ] extends Cond ? K : never
	};

export type FilterNames< T, Cond > = FilterProperties< T, Cond >[ keyof T ];

export type SubType< T, Cond, Invert = false > =
	Invert extends true
	? Omit< T, FilterNames< T, Cond > >
	: Pick< T, FilterNames< T, Cond > >;

export abstract class RecursiveValue { }
