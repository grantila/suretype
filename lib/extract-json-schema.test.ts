import {
	extractJsonSchema,
	extractSingleJsonSchema,
} from "./extract-json-schema.js"
import { suretype, v } from "./api/index.js"
import { ExportRefMethod } from "./types.js"


describe( "extract-json-schema", ( ) =>
{
	it( "suretype() should keep validator schema", ( ) =>
	{
		const schema = v.object( {
			foo: v.string( ).const( "bar" ).optional( ),
			bar: v.number( ).gt( 17 ),
		} );
		const schema2 = suretype( { name: "Foo" }, schema );

		expect(
			extractSingleJsonSchema( schema ).schema
		).toStrictEqual(
			extractSingleJsonSchema( schema2 ).schema
		);
		expect( schema.constructor ).toEqual( schema2.constructor );
	} );

	it( "suretype() of an array should produce valid JSON schema", ( ) =>
	{
		const inner = v.object( {
			foo: v.string( ).const( "bar" ).optional( ),
			bar: v.number( ).gt( 17 ),
		} );
		const schema = suretype(
			{
				name: "Foo",
				description: "Description",
				title: "Title",
				examples: [ "Example" ],
			},
			inner
		);

		const { schema: jsonSchemaInner } = extractSingleJsonSchema( inner );
		const { schema: jsonSchemaOuter } = extractJsonSchema( [ schema ] );

		expect( jsonSchemaOuter ).toEqual( {
			definitions: {
				Foo: {
					description: "Description",
					title: "Title",
					examples: [ "Example" ],
					...jsonSchemaInner,
				}
			}
		} );
	} );

	it( "should produce valid JSON schema for a 2+sized array", ( ) =>
	{
		const inner1 = v.object( {
			foo: v.string( ).const( "bar" ).optional( ),
		} );
		const inner2 = v.object( {
			bar: v.number( ).gt( 17 ),
		} );
		const schema1 = suretype(
			{
				name: "Foo",
				description: "Description",
				title: "Title",
				examples: [ "Example" ],
			},
			inner1
		);
		const schema2 = suretype(
			{
				name: "Bar",
				description: "Desc",
				title: "T",
				examples: [ "Ex" ],
			},
			inner2
		);

		const { schema: jsonSchemaInner1 } = extractSingleJsonSchema( inner1 );
		const { schema: jsonSchemaInner2 } = extractSingleJsonSchema( inner2 );
		const { schema: jsonSchemaOuter } =
			extractJsonSchema( [ schema1, schema2 ] );

		expect( jsonSchemaOuter ).toEqual( {
			definitions: {
				Foo: {
					description: "Description",
					title: "Title",
					examples: [ "Example" ],
					...jsonSchemaInner1,
				},
				Bar: {
					description: "Desc",
					title: "T",
					examples: [ "Ex" ],
					...jsonSchemaInner2,
				},
			}
		} );
	} );

	it( "should fail on non-decorated top-level validator schemas", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = v.string( );

		expect( ( ) => extractJsonSchema( [ schema1, schema2 ] ) )
			.toThrowError( /unnamed/ );
	} );

	it( "should ignore non-decorated validator schemas", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = v.string( );

		const { schema } = extractJsonSchema(
			[ schema1, schema2 ],
			{ onNonSuretypeValidator: 'ignore' }
		);

		expect( schema ).toMatchSnapshot( );
	} );

	it( "should call non-decorated validator schemas 'Unknown'", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = v.string( );

		const { schema } = extractJsonSchema(
			[ schema1, schema2 ],
			{ onNonSuretypeValidator: 'create-name' }
		);

		expect( schema ).toMatchSnapshot( );
	} );

	it( "should fail on duplicate top-level names", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = suretype(
			{ name: "Foo" },
			v.string( )
		);

		expect( ( ) => extractJsonSchema( [ schema1, schema2 ] ) )
			.toThrow( /uplicate.*Foo/ );
	} );

	it( "should be able to rename top-level names", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = suretype(
			{ name: "Foo" },
			v.string( )
		);

		const { schema } = extractJsonSchema(
			[ schema1, schema2 ],
			{ onTopLevelNameConflict: 'rename' }
		);

		expect( schema ).toMatchSnapshot( );
	} );

	it( "should produce two (referenced) definitions for 2 schemas", ( ) =>
	{
		const schema1 = suretype(
			{
				name: "Foo",
				description: "Description",
				title: "Title",
				examples: [ "Example" ],
			},
			v.object( { foo: v.string( ).const( "bar" ).optional( ) } )
		);
		const schema2 = suretype(
			{
				name: "Bar",
				description: "Desc",
				title: "T",
				examples: [ "Ex" ],
			},
			v.object( {
				bar: v.number( ).gt( 17 ),
				foo: schema1.optional( ),
			} )
		);

		const { schema: jsonSchemaOuter } =
			extractJsonSchema( [ schema1, schema2 ] );

		expect( jsonSchemaOuter ).toMatchSnapshot( );
	} );

	const methods: Array< ExportRefMethod > =
		[ 'no-refs', 'provided', 'ref-all' ];

	for ( const refMethod of methods )
	{
		it(
			"should produce proper definitions for referenced schemas " +
			`(using method ${refMethod})`,
			( ) =>
		{
			const inlined = suretype(
				{
					name: 'Baz'
				},
				v.number( )
			);
			const schema1 = suretype(
				{
					name: "Foo",
					description: "Description",
					title: "Title",
					examples: [ "Example" ],
				},
				v.object( { foo: v.string( ).const( "bar" ).optional( ) } )
			);
			const schema2 = suretype(
				{
					name: "Bar",
					description: "Desc",
					title: "T",
					examples: [ "Ex" ],
				},
				v.object( {
					bar: v.number( ).gt( 17 ),
					foo: schema1.optional( ),
					baz: inlined.optional( ),
					bak: inlined.optional( ),
				} )
			);

			const { schema: jsonSchemaOuter } =
				extractJsonSchema( [ schema1, schema2 ], { refMethod } );

			expect( jsonSchemaOuter ).toMatchSnapshot( );
		} )
	}

	it( "should keep references in lookup table", ( ) =>
	{
		const schema1 = suretype(
			{ name: "Foo" },
			v.number( )
		);
		const schema2 = v.anyOf( [ v.string( ), v.boolean( ) ] );
		const schema3 = v.anyOf( [ v.string( ), schema1 ] );

		const { schema, lookup, schemaRefName } = extractJsonSchema(
			[ schema1, schema2, schema3 ],
			{ onNonSuretypeValidator: 'lookup' }
		);

		expect( schema ).toMatchSnapshot( );
		expect( lookup.get( schema1 ) ).toMatchSnapshot( );
		expect( lookup.get( schema2 ) ).toMatchSnapshot( );
		expect( lookup.get( schema3 ) ).toMatchSnapshot( );
		expect( schemaRefName.get( lookup.get( schema1 ) ) ).toBe( 'Foo' );
		expect( schemaRefName.get( lookup.get( schema2 ) ) ).toBeUndefined( );
		expect( schemaRefName.get( lookup.get( schema3 ) ) ).toBeUndefined( );
	} );
} );
