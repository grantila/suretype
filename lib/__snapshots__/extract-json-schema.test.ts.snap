// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-json-schema should be able to rename top-level names 1`] = `
{
  "definitions": {
    "Foo": {
      "type": "number",
    },
    "Foo_2": {
      "type": "string",
    },
  },
}
`;

exports[`extract-json-schema should call non-decorated validator schemas 'Unknown' 1`] = `
{
  "definitions": {
    "Foo": {
      "type": "number",
    },
    "Unknown_1": {
      "type": "string",
    },
  },
}
`;

exports[`extract-json-schema should ignore non-decorated validator schemas 1`] = `
{
  "definitions": {
    "Foo": {
      "type": "number",
    },
  },
}
`;

exports[`extract-json-schema should keep references in lookup table 1`] = `
{
  "definitions": {
    "Foo": {
      "type": "number",
    },
  },
}
`;

exports[`extract-json-schema should keep references in lookup table 2`] = `
{
  "type": "number",
}
`;

exports[`extract-json-schema should keep references in lookup table 3`] = `
{
  "anyOf": [
    {
      "type": "string",
    },
    {
      "type": "boolean",
    },
  ],
}
`;

exports[`extract-json-schema should keep references in lookup table 4`] = `
{
  "anyOf": [
    {
      "type": "string",
    },
    {
      "$ref": "#/definitions/Foo",
    },
  ],
}
`;

exports[`extract-json-schema should produce proper definitions for referenced schemas (using method no-refs) 1`] = `
{
  "definitions": {
    "Bar": {
      "description": "Desc",
      "examples": [
        "Ex",
      ],
      "properties": {
        "bak": {
          "type": "number",
        },
        "bar": {
          "exclusiveMinimum": 17,
          "type": "number",
        },
        "baz": {
          "type": "number",
        },
        "foo": {
          "description": "Description",
          "examples": [
            "Example",
          ],
          "properties": {
            "foo": {
              "const": "bar",
              "type": "string",
            },
          },
          "title": "Title",
          "type": "object",
        },
      },
      "required": [
        "bar",
      ],
      "title": "T",
      "type": "object",
    },
    "Foo": {
      "description": "Description",
      "examples": [
        "Example",
      ],
      "properties": {
        "foo": {
          "const": "bar",
          "type": "string",
        },
      },
      "title": "Title",
      "type": "object",
    },
  },
}
`;

exports[`extract-json-schema should produce proper definitions for referenced schemas (using method provided) 1`] = `
{
  "definitions": {
    "Bar": {
      "description": "Desc",
      "examples": [
        "Ex",
      ],
      "properties": {
        "bak": {
          "type": "number",
        },
        "bar": {
          "exclusiveMinimum": 17,
          "type": "number",
        },
        "baz": {
          "type": "number",
        },
        "foo": {
          "$ref": "#/definitions/Foo",
        },
      },
      "required": [
        "bar",
      ],
      "title": "T",
      "type": "object",
    },
    "Foo": {
      "description": "Description",
      "examples": [
        "Example",
      ],
      "properties": {
        "foo": {
          "const": "bar",
          "type": "string",
        },
      },
      "title": "Title",
      "type": "object",
    },
  },
}
`;

exports[`extract-json-schema should produce proper definitions for referenced schemas (using method ref-all) 1`] = `
{
  "definitions": {
    "Bar": {
      "description": "Desc",
      "examples": [
        "Ex",
      ],
      "properties": {
        "bak": {
          "$ref": "#/definitions/Baz",
        },
        "bar": {
          "exclusiveMinimum": 17,
          "type": "number",
        },
        "baz": {
          "$ref": "#/definitions/Baz",
        },
        "foo": {
          "$ref": "#/definitions/Foo",
        },
      },
      "required": [
        "bar",
      ],
      "title": "T",
      "type": "object",
    },
    "Baz": {
      "type": "number",
    },
    "Foo": {
      "description": "Description",
      "examples": [
        "Example",
      ],
      "properties": {
        "foo": {
          "const": "bar",
          "type": "string",
        },
      },
      "title": "Title",
      "type": "object",
    },
  },
}
`;

exports[`extract-json-schema should produce two (referenced) definitions for 2 schemas 1`] = `
{
  "definitions": {
    "Bar": {
      "description": "Desc",
      "examples": [
        "Ex",
      ],
      "properties": {
        "bar": {
          "exclusiveMinimum": 17,
          "type": "number",
        },
        "foo": {
          "$ref": "#/definitions/Foo",
        },
      },
      "required": [
        "bar",
      ],
      "title": "T",
      "type": "object",
    },
    "Foo": {
      "description": "Description",
      "examples": [
        "Example",
      ],
      "properties": {
        "foo": {
          "const": "bar",
          "type": "string",
        },
      },
      "title": "Title",
      "type": "object",
    },
  },
}
`;
